Store average accelerometer readings in Object Model

BrendonBuilds opened the pull request:
Store the average value of each axis of the last accelerometer run in the Object Model

BrendonBuilds committed:
Store average accelerometer readings in Object Model

Store the average value of each axis of the last accelerometer run in the Object Model

BrendonBuilds commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -68,6 +68,8 @@ static volatile uint32_t numSamplesRequested;
 static uint8_t resolution = DefaultResolution;
 static uint8_t orientation = 20;							// +Z -> +Z, +X -> +X
 static volatile uint8_t axesRequested;
+constexpr size_t NumAxis = 3;
```
Is this var name too generic?

dc42 replied:
NumAccelerometerAxes would be better

BrendonBuilds commented on src/CAN/ExpansionManager.cpp:
```diff
@@ -24,6 +24,18 @@
+{
+	{
+		nullptr,					// no lock needed
+		[] (const ObjectModel *self, const ObjectExplorationContext& context) noexcept -> size_t { return 3; },
```
How would you prefer array sizes are tracked?

dc42 replied:
Move NumAccelerometerAxes (renamed from NumAxis) into an existing common header file so that you can use that value instead of 3.

BrendonBuilds commented on src/Platform/Platform.cpp:
```diff
@@ -219,6 +219,13 @@ constexpr ObjectModelArrayTableEntry Platform::objectModelArrayTable[] =
+	// 2. Average readings from last accelerometer run
+	{
+		nullptr,					// no lock needed
+		[] (const ObjectModel *self, const ObjectExplorationContext& context) noexcept -> size_t { return 3; },
```
How would you prefer array sizes are tracked?

dc42 replied:
See earlier comment

BrendonBuilds commented on src/CAN/ExpansionManager.h:
```diff
@@ -26,6 +26,7 @@ struct ExpansionBoardData
 
 	const char *_ecv_array typeName;
 	MinCurMax mcuTemp, vin, v12;
+	float accelerometerLastRunAverages[3];
```
How would you prefer array sizes are defined?

dc42 replied:
See previous comment

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -87,6 +89,15 @@ static void AddLocalAccelerometerRun(unsigned int numDataPoints) noexcept
 	reprap.BoardsUpdated();
 }
 
+static void AddLocalAccelerometerAverages(float averages[]) noexcept
```
Why not add extra arguments to AddLocalAccelerometerRun instead of adding this function?

BrendonBuilds replied:
I will do the same thing for `AddAccelerometerRun` as well

BrendonBuilds committed:
Adjustments to accelerometer averaging

Add extra arguments to AddLocalAccelerometerRun instead of a separate function

Move NumAccelerometerAxes to CANLib\RRF3Common.h

BrendonBuilds committed:
Merge branch 'Duet3D:3.5-dev' into 3.5-dev

BrendonBuilds commented:
The second commit will require the changes in: https://github.com/Duet3D/CANlib/pull/8

BrendonBuilds committed:
Merge branch '3.5-dev' of https://github.com/BrendonBuilds/RepRapFirmware into 3.5-dev

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -82,26 +82,23 @@ static IoPort spiCsPort;
-static void AddLocalAccelerometerAverages(float averages[]) noexcept
-{
-	for(unsigned int axis = 0;axis < NumAxis;++axis)
+	for(unsigned int axis = 0;axis < NumAccelerometerAxes;++axis)
```
Suggest use memcpyf here instead of a loop

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -479,12 +477,12 @@ GCodeResult Accelerometers::StartAccelerometer(GCodeBuffer& gb, const StringRef&
 		if (device.IsRemote())
 		{
-			reprap.GetExpansion().AddAccelerometerRun(device.boardAddress, 0);
+			reprap.GetExpansion().AddAccelerometerRun(device.boardAddress, 0, {0});
```
Need to pass an array of the correct size here because the compiler doesn't know how many values are expected

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -479,12 +477,12 @@ GCodeResult Accelerometers::StartAccelerometer(GCodeBuffer& gb, const StringRef&
 # endif
 		{
-			AddLocalAccelerometerRun(0);
+			AddLocalAccelerometerRun(0, {0});
```
Need to pass an array of the correct size here because the compiler doesn't know how many values are expected, or add a new function AddFailedLocalAccelerometerRun that doesn't take the averages parameter

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -515,7 +513,7 @@ GCodeResult Accelerometers::StartAccelerometer(GCodeBuffer& gb, const StringRef&
 			accelerometerFile = nullptr;
 			MassStorage::Delete(accelerometerFileName.c_str(), false);
-			reprap.GetExpansion().AddAccelerometerRun(device.boardAddress, 0);
+			reprap.GetExpansion().AddAccelerometerRun(device.boardAddress, 0, {0});
```
Need to pass an array of the correct size here because the compiler doesn't know how many values are expected. Alternatively, add another function AddFailedAccelerometerRun  that doesn't take the averages as a parameter and instead sets them to zero, since this appears to be done in several places.

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -599,15 +597,15 @@ void Accelerometers::ProcessReceivedData(CanAddress src, const CanMessageAcceler
 			f->Close();
 			accelerometerFile = nullptr;
-			reprap.GetExpansion().AddAccelerometerRun(src, 0);
+			reprap.GetExpansion().AddAccelerometerRun(src, 0, {0});
```
See previous comment.

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -599,15 +597,15 @@ void Accelerometers::ProcessReceivedData(CanAddress src, const CanMessageAcceler
 			f->Close();
 			accelerometerFile = nullptr;
-			reprap.GetExpansion().AddAccelerometerRun(src, 0);
+			reprap.GetExpansion().AddAccelerometerRun(src, 0, {0});
```
See previous comment.

dc42 commented on src/Accelerometers/Accelerometers.cpp:
```diff
@@ -679,14 +677,14 @@ void Accelerometers::ProcessReceivedData(CanAddress src, const CanMessageAcceler
 
 				// find the average value for each axis
-				for (unsigned int axis = 0; axis < NumAxis; ++axis)
+				for (unsigned int axis = 0; axis < NumAccelerometerAxes; ++axis)
```
Better to use a range-based for-loop here

dc42 commented on src/CAN/ExpansionManager.cpp:
```diff
@@ -307,19 +307,16 @@ void ExpansionManager::UpdateFailed(CanAddress address) noexcept
+	boards[address].accelerometerLastRunDataPoints = numDataPoints;
+	++boards[address].accelerometerRuns;
+
+	for(int32_t i = 0;i < NumAccelerometerAxes;++i)
```
Could we use memcpyf here?

BrendonBuilds committed:
Add Add(Local)FailedAccelerometerRun

BrendonBuilds committed:
fix compiler warnings

T3P3 commented:
We are closing all old PRS that don't target 3.7-dev branch
